pkg/workflow/workflow.go

```diff
@@ -2,25 +2,216 @@ package workflow
 
 import (
 	"context"
+	"fmt"
 
-	"github.com/Azure/Orkestra/api/v1alpha1"
 	v1alpha12 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Azure/Orkestra/api/v1alpha1"
 	"github.com/go-logr/logr"
 )
 
+type ClientType string
+
+type ExecutorFunc func(string, ExecutorAction) v1alpha12.Template
+
 const (
-	Project        = "orkestra"
-	OwnershipLabel = "owner"
-	HeritageLabel  = "heritage"
+	Forward ClientType = "forward"
+	Reverse ClientType = "reverse"
 )
 
-type Engine interface {
+var _ = ForwardWorkflowClient{}
+var _ = ReverseWorkflowClient{}
+
+type Client interface {
 	// Generate the object required by the workflow engine
-	Generate(ctx context.Context, l logr.Logger, g *v1alpha1.ApplicationGroup) error
+	Generate() error
 	// Submit the object required by the workflow engine generated by the Generate method
-	Submit(ctx context.Context, l logr.Logger, g *v1alpha1.ApplicationGroup) error
-	// Generate the reverse object required by the workflow engine
-	GenerateReverse(ctx context.Context, l logr.Logger, nodes map[string]v1alpha12.NodeStatus, wf *v1alpha12.Workflow) error
-	// Submit the reverse object required by the workflow engine generated by the GenerateReverse method
-	SubmitReverse(ctx context.Context, l logr.Logger, wf *v1alpha12.Workflow) error
+	Submit(ctx context.Context) error
+
+	GetLogger() logr.Logger
+
+	GetWorkflow(context.Context) (*v1alpha12.Workflow, error)
+
+	GetClient() client.Client
+}
+
+type ClientOptions struct {
+	parallelism *int64
+	stagingRepo string
+	namespace   string
+}
+
+type Builder struct {
+	client     client.Client
+	logger     logr.Logger
+	clientType ClientType
+	options    ClientOptions
+	executor   ExecutorFunc
```

Comment:
This assumes both forward and reverse executors are the same - can we provide an option to configure them individually.

Reply:
It doesn't necessarily. You can create a "new builder" with new options if you want to create a client with different options